### ISSUE TITLE
[2.9.1-rc] Why was a -rc1 version pushed to npm registry as latest version?

### ISSUE THREAD

**Issue by triccardi-systran** [jshint/jshint#2765]
Hi,

Why was a RC version pushed to the npm registry as `lastest` version, and not with a special dist-tag ?

`npm view jshint` extract:
```
  'dist-tags': { latest: '2.9.1-rc1' },
```

**Comment by jugglinmike** [comment]
Hey there! I published that release, but I wasn't aware of distribution tags. If you share a little context about why you are asking, then I may be able to help.

**Comment by triccardi-systran** [comment]
Our build run gulp-jshint, and recently got broken because of new warnings appearing, for example:
`line 129  col 7   'dataTable' was used before it was defined.` (all examples seem to be about named functions).
When investigating it appeared that with previous `jshint` 2.8.0 those warnings weren't there.

When finding that the current `latest` vestion on npm (ie the one installed when running `npm install jshint`) was a RC version, I opened this issue as it doesn't seem right to me to have a RC version as the standard one.

Afterward I found in the changelog:
> In an effort to minimize friction with a new version, we're publishing a release candidate and requesting feedback from early adopters. Please give it a try in your projects and let us know about any surprising behavior!

For me this means this version is not OK for production yet, thus the RC version, but this also implies not pushing it as `latest` version on npm, doesn't it?


What I would expect: 
* either it's a ready for prod final version, then it should not have a "-rc1" suffix in version
* or it's a RC version, and it should be pushed in a dist-tag different than `latest` (maybe `rc`, I don't know the best practices for that), with documentation explaining how to install this rc version for `early adopters`.

**Comment by triccardi-systran** [comment]
Okay, so the culprid is `node.js` v0.10.40 (latest v0.10.x), which ships the old `npm` v1.4.28, which itself ships the old `semver` v2.3.0:
```
$ npm install semver@2.3.0
semver@2.3.0 node_modules/semver
$ node
> var semver = require('semver')
undefined
> semver.satisfies('2.9.1-rc1', '^2.7.0')
true
```

It also happens with `semver` 2.3.2 (latest 2.3.x), 3.0.1 (latest 3.x).
The behavior first changed in `semver` 4.0.0.

`npm` 1.4.29 that will ship in next `node.js` v0.10 will include a warning to recommend upgrading to npm >= 2.x, which has changes in semver (and those changes are probably ours): https://github.com/npm/npm/releases/tag/v1.4.29


We will upgrade npm on our build systems.
However, the distribution tags may still be useful for such use-case. Maybe ask the community?

**Comment by lukeapage** [comment]
the trouble with distribution tags is what to do when a rc is finished.. you end up having to publish and maintain each tag. e.g. say the next release didn't warrant a rc, you'd still have to publish it under that tag also for people using rc as a kind-of "latest".
IMO semver is a better solution, but unfortunate for you that it wasn't supported in old npm.

However..

> Our build run gulp-jshint, and recently got broken because of new warnings appearing, for example:
> line 129 col 7 'dataTable' was used before it was defined. (all examples seem to be about named functions).
> When inves

this is probably a valid warning. what do you do when there is a bug in jshint that prevents a warning which under the spirit of the rule should actually be an error? We did go up a minor version (not just patch) to introduce this.

If you post some code in a separate issue and cc me I'll tell you whether its a regression, but I doubt it.

**Comment by triccardi-systran** [comment]
The `rc` tag could be a one shot, just when needed, but it would be indeed harder to explain to the users (Maybe something like that: "This dist tag is for the latest `rc` version, which don't always exist for all future versions of `jshint`: don't use this to always get the ~next version.").
I understand it's not perfect.

However, `npm` 1.4.28 is still used by many people because it's shipped with the latest node.js 0.10. Maybe add a big warning on your README? and on all dependent npm packages README? Not great either...
Future node.js 0.10 version will have a warning when running npm 1.4.29, which hopefully will make users migrate to npm 2 or more. I don't know when it will be released though...


As for the initial issue I encountered, it's an issue in our code: jshint just correctly found more warnings. I explained the whole situation simply because @jugglinmike asked for it.
The new warnings situation just didn't happen to us before.


To conclude, When the new node.js 0.10 version will be released (assuming it'll ship npm 1.4.29), this issue will be closed. In the meantime, maybe add a warning on your README?

**Comment by jugglinmike** [comment]
This feels like a somewhat fringe case to warrant mention in the project's README. I think it would be more fitting for http://jshint.com/docs/faq/ -- if you'd like to follow up with a patch to [the website repository](https://github.com/jshint/jshint.github.io), I'd be happy to review for you.

Either way, thanks for helping us get to the bottom of this!
